Lib/functools.py

```diff
@@ -274,43 +275,129 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
-# Purely functional, no descriptor behaviour
-class partial:
-    """New function with partial application of the given arguments
-    and keywords.
+
+class _PlaceholderType:
+    """The type of the Placeholder singleton.
+
+    Used as a placeholder for partial arguments.
     """
+    __instance = None
+    __slots__ = ()
 
-    __slots__ = "func", "args", "keywords", "__dict__", "__weakref__"
+    def __init_subclass__(cls, *args, **kwargs):
+        raise TypeError(f"type '{cls.__name__}' is not an acceptable base type")
 
-    def __new__(cls, func, /, *args, **keywords):
+    def __new__(cls):
+        if cls.__instance is None:
+            cls.__instance = object.__new__(cls)
+        return cls.__instance
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __reduce__(self):
+        return 'Placeholder'
+
+Placeholder = _PlaceholderType()
+
+def _partial_prepare_merger(args):
+    nargs = len(args)
+    if not nargs:
+        return 0, None
+    order = list()
+    i, j = 0, nargs
+    for a in args:
+        if a is Placeholder:
+            order.append(j)
+            j += 1
+        else:
+            order.append(i)
+        i += 1
+    phcount = j - nargs
+    merger = itemgetter(*order) if phcount else None
+    return phcount, merger
+
+def _partial_new(cls, func, /, *args, **keywords):
+    if issubclass(cls, partial):
         if not callable(func):
             raise TypeError("the first argument must be callable")
+    else:
+        # assert issubclass(cls, partialmethod)
+        if not callable(func) and not hasattr(func, "__get__"):
+            raise TypeError(f"{func!r} is not callable or a descriptor")
+            # func could be a descriptor like classmethod which isn't callable,
+            # so we can't inherit from partial (it verifies func is callable)
+            # flattening is mandatory in order to place cls/self before all
+            # other arguments
+            # it's also more efficient since only one function will be called
+    if args and args[-1] is Placeholder:
+        raise TypeError("trailing Placeholders are not allowed")
+    if isinstance(func, cls):
+        pto_phcount = func._phcount
+        tot_args = func.args
+        if args:
+            tot_args += args
+            if pto_phcount:
+                # merge args with args of `func` which is `partial`
+                nargs = len(args)
+                if nargs < pto_phcount:
+                    tot_args += (Placeholder,) * (pto_phcount - nargs)
+                tot_args = func._merger(tot_args)
+                if nargs > pto_phcount:
+                    tot_args += args[pto_phcount:]
+            phcount, merger = _partial_prepare_merger(tot_args)
+        elif pto_phcount:   # not args
```

Comment:
Is this test needed?

Reply:
I think so. It is a 2x2 conditional block on `args` and `pto_phcount`. This is one of the cases - `not args and pto_hcount`. Am I missing something?

Reply:
It looks to me that the `func._merger` is already None if `pto_phcount` is 0, so special case for this is not needed. Tests are passed if replace it with `else:`.

Reply:
That condition is for when inner `partial` has placeholders, but outer `partial` has no arguments. Then it simply inherits `_merger`.

I added a test which fails.

Reply:
Ah ok, I see. I think I did it that way because I thought that condition check is faster than attribute getter.

But I changed to your suggestion, simpler is better here